config/finder.go

```diff
@@ -63,46 +63,106 @@ func GetConfigFiles(targetDir string) ([]Config, error) {
 	currentDir := path.Dir(start)
 	emptyMap := make(map[string]string)
-
 	for i := 0; i < maxDepth; i++ {
-		configFileRelativePath := findFile(currentDir, configFileName)
-		configDirRelativePath := findFile(currentDir, configDirName)
-		configDirConfigFileRelativePath := findFile(currentDir, configDirName, configFileName)
-
-		nothingFound := configFileRelativePath == "" && configDirRelativePath == ""
-		onlyConfigFileFound := configFileRelativePath != "" && configDirRelativePath == ""
-		onlyConfigDirFound := configFileRelativePath == "" && configDirRelativePath != ""
-		bothConfigTypesFound := configFileRelativePath != "" && configDirRelativePath != ""
-		configDirMissingConfig := onlyConfigDirFound && configDirConfigFileRelativePath == ""
-
-		if bothConfigTypesFound {
-			return "", emptyMap, fmt.Errorf("in %s you must use either a config file or config directory but not both", currentDir)
-		}
-
-		if configDirMissingConfig {
-			return "", emptyMap, fmt.Errorf("config dir %s is missing the % file", configDirRelativePath, configFileName)
+		configStatus, configFileRelativePath, templateMap, err := collateConfig(start, configFileName, configDirName)
+		if err != nil {
+			return "", emptyMap, err
 		}
 
-		if nothingFound {
+		if configStatus == "NothingFound" {
 			currentDir = path.Join(currentDir, "..")
 			continue
+		} else {
+			return configFileRelativePath, templateMap, nil
 		}
+	}
+	return "", emptyMap, fmt.Errorf("root config file not found")
+}
 
-		if onlyConfigFileFound {
-			return configFileRelativePath, emptyMap, nil
-		}
+// Examine the TFGEN configuration situation in a specific directory
+func collateConfig(targetDir, configFileName, configDirName string) (string, string, map[string]string, error) {
 
-		if onlyConfigDirFound {
-			templateFiles, err := findTemplateFiles(configDirRelativePath, []string{configFileName})
-			if err != nil {
-				return "", emptyMap, err
-			}
-			return configDirConfigFileRelativePath, templateFiles, nil
+	currentDir := path.Dir(targetDir)
+	emptyMap := make(map[string]string)
+	configFileRelativePath := findFile(currentDir, configFileName)
+	configDirRelativePath := findFile(currentDir, configDirName)
+	configDirConfigFileRelativePath := findFile(currentDir, configDirName, configFileName)
+
+	nothingFound := configFileRelativePath == "" && configDirRelativePath == ""
+	onlyConfigFileFound := configFileRelativePath != "" && configDirRelativePath == ""
+	onlyConfigDirFound := configFileRelativePath == "" && configDirRelativePath != ""
+	bothConfigTypesFound := configFileRelativePath != "" && configDirRelativePath != ""
+	configDirMissingConfig := onlyConfigDirFound && configDirConfigFileRelativePath == ""
+
+	if bothConfigTypesFound {
+		return "BothConfigTypesFound", "", emptyMap, fmt.Errorf("in %s you must use either a config file or config directory but not both", currentDir)
```

Comment:
It occurs to me that these return values for config status might be nice to have as an enum or some other pattern where if they were misused somehow accidentally, it would show up easier with compilation errors etc.

Reply:
Agreed. I was under the impression enums did not exist in Go, but after some digging I've seen there are "official" ways to implement them. I think there would be some value updating these statuses accordingly at some point.